tools/themis/src/rules.rs

```diff
@@ -0,0 +1,329 @@
+use std::collections::HashMap;
+
+use super::{style, utils, Error, Expected, PackageOutcome, Workspace};
+
+/// ensure all crates have the `publish = <true/false>` specification
+pub fn has_publish_spec(workspace: &Workspace) -> Result<(), Error> {
+    let outliers: Vec<_> = workspace
+        .members
+        .iter()
+        .filter(|pkg| pkg.raw["package"].get("publish").is_none())
+        .map(|pkg| PackageOutcome { pkg, value: None })
+        .collect();
+
+    if !outliers.is_empty() {
+        return Err(Error::OutcomeError {
+            msg: "These packages should have the `publish` specification".to_string(),
+            expected: Some(Expected { value: "publish = <true/false>".to_string(), reason: None }),
+            outliers,
+        });
+    }
+
+    Ok(())
+}
```

Comment:
Theoretically, this can be written without if statements. I'm not saying we should, but just a theoretically experiment.

```
pub fn has_publish_spec(workspace: &Workspace) -> Result<(), Error> {
    once(
        workspace
            .members
            .iter()
            .filter(|pkg| pkg.raw["package"].get("publish").is_none())
            .map(|pkg| PackageOutcome { pkg, value: None })
            .collect::<Vec<_>>(),
    )
    .into_iter()
    .filter(|outliers| !outliers.is_empty())
    .map(|outliers| Error::OutcomeError {
        msg: "These packages should have the `publish` specification".to_string(),
        expected: Some(Expected { value: "publish = <true/false>".to_string(), reason: None }),
        outliers,
    })
    .next()
    .map(|e| Err(e))
    .unwrap_or(Ok(()))
}
```

Reply:
@matklad I wish there was a nicer API for doing things like this.

Reply:
Instead of ```
    .next()
    .unwrap_or(Ok(()))
```
something like
```
    .next_or(Ok(()))
```

Reply:
I wish I could write it like
```
    workspace
        .members
        .iter()
        .filter(|pkg| pkg.raw["package"].get("publish").is_none())
        .map(|pkg| PackageOutcome { pkg, value: None })
        .collect::<Vec<_>>()
    .once()
    .filter(|outliers| !outliers.is_empty())
    .map(|outliers| {
        Err(Error::OutcomeError {
            msg: "These packages should have the `publish` specification".to_string(),
            expected: Some(Expected { value: "publish = <true/false>".to_string(), reason: None }),
            outliers,
        })
    })
    .next_or(Ok(()))
```

or even better


```
    workspace
        .members
        .iter()
        .filter(|pkg| pkg.raw["package"].get("publish").is_none())
        .map(|pkg| PackageOutcome { pkg, value: None })
    .once()
    .filter(|outliers| !outliers.is_empty())
    .map(|outliers| {
        Err(Error::OutcomeError {
            msg: "These packages should have the `publish` specification".to_string(),
            expected: Some(Expected { value: "publish = <true/false>".to_string(), reason: None }),
            outliers: outliers.collect(),
        })
    })
    .next_or(Ok(()))
```

Reply:
> We don't have `is_empty` on `iter()`

While we actually do have [`is_empty`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#method.is_empty) on structs implementing `ExactSizeIterator`, in this case, we don't know how many times this would iterate. So this is not an `ExactSizeIterator`, which means the only way to check `is_empty` would be to consume the first item.

The logic as defined is evidently the clearer way to write this.

Reply:
> We don't have `is_empty` on `iter()`

While we actually do have [`is_empty`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#method.is_empty) on structs implementing `ExactSizeIterator`, in this case, we don't know how many times this would iterate. So this is not an `ExactSizeIterator`, which means the only way to check `is_empty` would be to consume the first item.

The logic as defined is evidently the clearer way to write this.

Reply:
Yes, I agree.

Another way, to write the logic would be to write it avoiding `return` statement.

```
    if !outliers.is_empty() {
        return Err(Error::OutcomeError {
            msg: "These packages shouldn't be versioned".to_string(),
            expected: Some(Expected { value: "0.0.0".to_string(), reason: None }),
            outliers,
        });
    }
    OK(())
```
would become
```
    if outliers.is_empty() {
        OK(())
    } else {
         Err(Error::OutcomeError {
            msg: "These packages shouldn't be versioned".to_string(),
            expected: Some(Expected { value: "0.0.0".to_string(), reason: None }),
            outliers,
        });
    }
```

Reply:
@pmnoxx I think you are looking for find_map: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.find_map

Though, in this case this won't work -- we indeed want to collect all of the results, instead of just the first one.